tokio/src/sync/watch.rs

```diff
@@ -844,7 +845,9 @@ impl<T> Receiver<T> {
             }
 
             // Wait for the value to change.
-            closed = changed_impl(&self.shared, &mut self.version).await.is_err();
+            closed = cooperative(changed_impl(&self.shared, &mut self.version))
+                .await
+                .is_err();
```

Comment:
This doesn't make `wait_for` cooperative, as it may exit before it reaches this call without touching the budget. You need to wrap the entire function.

Also, you may want to make a `wait_for_inner` method instead of wrapping the entire body in `cooperative(async { ... })` to avoid extra indentation.

Reply:
Oh yea I totally missed that! Thanks. Done in [d5a6b3b](https://github.com/tokio-rs/tokio/pull/6846/commits/d5a6b3b7ca753e26c760d9af74e3e206edf97056).